### DISCUSSION TITLE
Context em ViewModel

### DISCUSSION THREAD

**Discussion by DIMAAGR** [Flutterando/forum#37] [Perguntas e Respostas] [answered]
Opa galera tudo beleza? Bom, estou aqui mais uma vez e dessa vez é para entender um pouco mais sobre o Context e o viewmodel, como o viewmodel faz as ações na view, seria correto passar o contexto para ela, não como um contexto global, mas como por exemplo como um parâmetro em uma função, se alguém puder tirar essa dúvida ficaria agradecido!

**Comment by JoaoWitorFelipe** [comment, marked as answer]
Eu imagino que o ViewModel seria o equivalente a um controlador (Controller, na ideia de MVC, mas me corrija se estiver errado). Particularmente, não acho interessante passar o context para uma camada que lida com gerência de estado ou que se responsabiliza em chamar serviços externos. Ao invés de passar o context para executar no viewmodel, talvez seja mais interessante trabalhar orientado a callbacks.

No seu ViewModel ao invés de fazer isso:

```dart

class ViewModel {
   Future<void> function(Context context) {
       // ...
   }
}
```
Eu faria isso, por exemplo:
```dart

class ViewModel {
   Future<void> function(VoidCallback onSuccess, VoidCallback onError) {
       // ...
   }
}
```
E na minha view, utilizar o context de acordo com os callback. 
Espero que tenha ajudado :smile:.

**Reply by DIMAAGR** [reply]
Muito obrigado pela resposta!!

**Comment by brunosemfio** [comment]
A princípio não acho que seja uma boa usar o contexto fora da view. Tem algum exemplo fácil ai?

**Reply by mateusfccp** [reply]
Sim. Via de regra, não é uma boa ideia ficar passando contexto de um lado pro outro.

Tudo que envolve contexto deve ser feito dentro da view.

O que não ficou claro para mim, é o que é esse ViewModel, já que esse nome não é unívoco. Por exemplo, se estivermos falando de Clean Architecture, o ViewModel não faz nada na view, então não faria sentido a pergunta.

Seria bom especificar de qual abordagem se refere a pergunta.

**Reply by DIMAAGR** [reply]
o viewmodel do mvvm
